test/python/algorithms/test_grover.py

```diff
@@ -230,40 +310,76 @@ def test_construct_circuit(self):
+            grover = Grover(iterations=[1, 2, 3, 4], sampler=self._sampler)
+        elif use_sampler == "shots":
+            grover = Grover(iterations=[1, 2, 3, 4], sampler=self._sampler_with_shots)
+        else:
+            with self.assertWarns(PendingDeprecationWarning):
+                grover = Grover(iterations=[1, 2, 3, 4], quantum_instance=self.qasm)
         result = grover.amplify(problem)
-        expected_results = [
-            {"11": 1024},
-            {"00": 238, "01": 253, "10": 263, "11": 270},
-            {"00": 238, "01": 253, "10": 263, "11": 270},
-            {"11": 1024},
-        ]
-        self.assertEqual(result.circuit_results, expected_results)
-
-    def test_max_probability(self):
+        if use_sampler:
+            for i, dist in enumerate(result.circuit_results):
+                keys, values = zip(*sorted(dist.items()))
+                self.assertTupleEqual(keys, ("00", "01", "10", "11"))
+                if i in (0, 3):
+                    np.testing.assert_allclose(values, [0, 0, 0, 1], atol=0.2)
+                else:
+                    np.testing.assert_allclose(values, [0.25, 0.25, 0.25, 0.25], atol=0.2)
+        else:
+            expected_results = [
+                {"11": 1024},
+                {"00": 238, "01": 253, "10": 263, "11": 270},
+                {"00": 238, "01": 253, "10": 263, "11": 270},
+                {"11": 1024},
+            ]
+            self.assertEqual(result.circuit_results, expected_results)
+
+    @data("ideal", "shots", False)
+    def test_max_probability(self, use_sampler):
         """Test max_probability"""
         oracle = QuantumCircuit(2)
         oracle.cz(0, 1)
         problem = AmplificationProblem(oracle, is_good_state=["11"])
-        grover = Grover(quantum_instance=self.qasm)
+        if use_sampler == "ideal":
+            grover = Grover(sampler=self._sampler)
+        elif use_sampler == "shots":
+            grover = Grover(sampler=self._sampler_with_shots)
+        else:
+            with self.assertWarns(PendingDeprecationWarning):
+                grover = Grover(quantum_instance=self.qasm)
         result = grover.amplify(problem)
-        self.assertEqual(result.max_probability, 1.0)
+        self.assertAlmostEqual(result.max_probability, 1.0)
 
-    def test_oracle_evaluation(self):
+    @data("ideal", "shots", False)
+    def test_oracle_evaluation(self, use_sampler):
         """Test oracle_evaluation for PhaseOracle"""
         oracle = PhaseOracle("x1 & x2 & (not x3)")
         problem = AmplificationProblem(oracle, is_good_state=oracle.evaluate_bitstring)
-        grover = Grover(quantum_instance=self.qasm)
+        if use_sampler == "ideal":
+            grover = Grover(sampler=self._sampler)
+        elif use_sampler == "shots":
+            grover = Grover(sampler=self._sampler_with_shots)
+        else:
+            with self.assertWarns(PendingDeprecationWarning):
+                grover = Grover(quantum_instance=self.qasm)
```

Comment:
This fragments looks redundant across many tests.

Reply:
Ya? It checks an ideal sampler, a sampler with shots, and a quantum instance. I added them based on the previous comments. Do you have any suggestions?

Reply:
I suggest extracting it as a method and reusing.

Reply:
It's much better now isn't it? 🙂

Reply:
Perfect! Thanks.